src/systems/mecanum_drive/MecanumDrive.cc

```diff
@@ -0,0 +1,719 @@
+  public: std::mutex mutex;
+
+  /// \brief frame_id from sdf.
+  public: std::string sdfFrameId;
+
+  /// \brief child_frame_id from sdf.
+  public: std::string sdfChildFrameId;
+};
+
+//////////////////////////////////////////////////
+MecanumDrive::MecanumDrive()
+  : dataPtr(std::make_unique<MecanumDrivePrivate>())
+{
+}
+
+//////////////////////////////////////////////////
+void MecanumDrive::Configure(const Entity &_entity,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    EntityComponentManager &_ecm,
+    EventManager &/*_eventMgr*/)
+{
+  this->dataPtr->model = Model(_entity);
+
+  // Get the canonical link
+  std::vector<Entity> links = _ecm.ChildrenByComponents(
+      this->dataPtr->model.Entity(), components::CanonicalLink());
+  if (!links.empty())
+    this->dataPtr->canonicalLink = Link(links[0]);
+
+  if (!this->dataPtr->model.Valid(_ecm))
+  {
+    ignerr << "MecanumDrive plugin should be attached to a model entity. "
+           << "Failed to initialize." << std::endl;
+    return;
+  }
+
+  // Ugly, but needed because the sdf::Element::GetElement is not a const
+  // function and _sdf is a const shared pointer to a const sdf::Element.
+  auto ptr = const_cast<sdf::Element *>(_sdf.get());
+
+  // Get params from SDF
+  sdf::ElementPtr sdfElem = ptr->GetElement("front_left_joint");
+  while (sdfElem)
+  {
+    this->dataPtr->frontLeftJointNames.push_back(sdfElem->Get<std::string>());
+    sdfElem = sdfElem->GetNextElement("front_left_joint");
+  }
+  sdfElem = ptr->GetElement("front_right_joint");
+  while (sdfElem)
+  {
+    this->dataPtr->frontRightJointNames.push_back(sdfElem->Get<std::string>());
+    sdfElem = sdfElem->GetNextElement("front_right_joint");
+  }
+
+  sdfElem = ptr->GetElement("back_left_joint");
+  while (sdfElem)
+  {
+    this->dataPtr->backLeftJointNames.push_back(sdfElem->Get<std::string>());
+    sdfElem = sdfElem->GetNextElement("back_left_joint");
+  }
+  sdfElem = ptr->GetElement("back_right_joint");
+  while (sdfElem)
+  {
+    this->dataPtr->backRightJointNames.push_back(sdfElem->Get<std::string>());
+    sdfElem = sdfElem->GetNextElement("back_right_joint");
+  }
+
+  this->dataPtr->wheelSeparation = _sdf->Get<double>("wheel_separation",
+      this->dataPtr->wheelSeparation).first;
+  this->dataPtr->wheelbase = _sdf->Get<double>("wheelbase",
+      this->dataPtr->wheelbase).first;
+  this->dataPtr->wheelRadius = _sdf->Get<double>("wheel_radius",
+      this->dataPtr->wheelRadius).first;
+
+  // Instantiate the speed limiters.
+  this->dataPtr->limiterLin = std::make_unique<ignition::math::SpeedLimiter>();
+  this->dataPtr->limiterAng = std::make_unique<ignition::math::SpeedLimiter>();
+
+  // Parse speed limiter parameters.
+  if (_sdf->HasElement("min_velocity"))
+  {
+    double minVel = _sdf->Get<double>("min_velocity");
+    this->dataPtr->limiterLin->SetMinVelocity(minVel);
+    this->dataPtr->limiterAng->SetMinVelocity(minVel);
```

Comment:
I realize that this is just following the pattern from other systems, but I'm wondering if we should split linear and angular components, considering that they have different units.